### ISSUE TITLE
Web: Make AttrBuilderContext Composable

### ISSUE THREAD

**Issue by hfhbd** [JetBrains/compose-jb#2304]
https://github.com/JetBrains/compose-jb/blob/471d5c652db20286d478c25c2b3d80ab4200215a/web/core/src/jsMain/kotlin/org/jetbrains/compose/web/elements/Base.kt#L21 It is already Composable, but this feature/annotation is no forward to other calls, namely: https://github.com/JetBrains/compose-jb/blob/471d5c652db20286d478c25c2b3d80ab4200215a/web/core/src/jsMain/kotlin/org/jetbrains/compose/web/elements/Base.kt#L109

Use case:
Use `remember` in the styling api to create ids automatically.

**Comment by eymar** [comment]
Could you please add a snippet how you would use `remember` in style or attrs blocks?

**Comment by hfhbd** [comment]
@eymar What the test of the PR? :D https://github.com/JetBrains/compose-jb/pull/2308/files#diff-96fe1b0a96b139b5b9b4fd22973cc6e48a52b8eb3fd477296bc29cb40b7b85b8

**Comment by hfhbd** [comment]
I did try to make `style` composable too, but the non composable test `attrsBuilderCopyFromPreservesExistingAttrs` failed (or my setup is somehow broken...)

**Comment by eymar** [comment]
```kotlin
Div({
                val color by remember {
                    derivedStateOf {
                        if (colorFlip) {
                            Color.green
                        } else Color.red
                    }
                }
                style {
                    backgroundColor(color)
                }
            })
        }
```

Perhaps, I'm missing something in this snippet. As I understand it now, we can move `val color by remember ...` above the Div, and use `color` in `style {}` and it should work. 
Maybe this test is just to validate that composable attrs block works. 
So maybe, your use-case `Use remember in the styling api to create ids automatically.` could help me understand your goal better. How would remember help create ids automatically?

**Comment by hfhbd** [comment]
@eymar Thanks for the feedback! The goal is to support `display-grid` (and some other css features): https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout
In the sample of MDN they use classes and bootstrap-compose generates these class names automatically with a UUID. So we use this:
```kotlin
Style {
    val classname = remember { "_${UUID()}" }

    className(classname) style {
        // apply the styling
    }
}
return classname
```
and the same class name in the actually html element.

But at the end, I dislike this behaviour. Instead I should use the `AttrsScope.style` function with "in-memory" style sheets instead writing the style in the DOM and use the same class name. 

While this PR would simplify my current implementation, I think, it would be better to implement it with `AttrsScope.style` instead, somehow in the long term.

So after rethinking the current bootstrap-compose implementation I would keep the behaviour in `compose-jb` and change my implementation instead.

**Comment by eymar** [comment]
To achieve what you need, would something like below snippet be helpful?

```kotlin
Div(attrs = {
    val id = composableCalculation { // Composable calls are allowed here
        return@composableCalculation remember { "someGeneratedId" }
    }
    id(id)
}) {
    Text("Hello World!")
}
```

`composableCalculation` could rely on a fact that in the end attrs are applied in Composable scope. It can be an unsafe assumption, because anyone can create AttrsScope instance and then apply attrs lambdas on it, and `composableCalculation` will fail with an error for self-instantiated AttrsScope.

**Comment by erikhuizinga** [comment]
If `AttrBuilderContext` would become `@Composable`, that would mean that it could call any and all composables, right? What is the expected behaviour if I call e.g the following?

```kotlin
Div(
    attrs = {
        Text("Hello, AttrBuilderContext!")
    }
) {}
```

Would / should that do
- nothing, or
- evaluate in-place and render that text at the current position in the HTML tree, or
- fail to compile, or
- crash at runtime, or
- something else?

**Comment by hfhbd** [comment]
@erikhuizinga Yeah, that's true and I think, we should not to enable `@Composable` in attrs scope because the usage is unambiguous, like you said. The option `composableCalculation` would be more explicit, but still unsafe. 

Ideally, the sample should not compile, but this would only work with some marker annotation. Even a `@DslMarker` is still unsafe, because you still could overwrite it with `attrs = { this.Text("") }`.

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-2304">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
